### DISCUSSION TITLE
Managing storage in Codespaces

### DISCUSSION THREAD

**Discussion by jkitchin** [community/community#19511] [Codespaces] [answered]
I have been using codespaces for a few weeks, and when I look in the Billing section, I can see the storage has creeped up to 9+ GB. It seems to only go up, and doesn't seem to go down when I delete codespaces. 

I am investigating using this for a class, and I imagine this will be 9GB * 50 before you know it!

Is there a way to manage this? e.g. to see what these are, and to delete things that are not needed?

**Comment by cdwort** [comment, marked as answer]
Hello @jkitchin Thanks for using Codespaces and this feedback!

Just to make sure I'm looking at the same view you are, is this example billing section (pulled from our [billing docs](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-codespaces)) what you're looking at?

![codespaces-compute-storage](https://user-images.githubusercontent.com/1014482/176377678-cc0a859e-5784-48ce-a7d7-da33399288da.png)

If so, the storage number you're seeing is actually GB/month (or, in most cases, GB/month pro-rated to however long the codespace existed during this month). For that reason, you won't see this number drop when you delete codespaces -- it's representing accrued usage. If you'd like to dig into your usage in detail, you can follow the final step in [these instructions](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/viewing-your-codespaces-usage) to pull a report.

That being said, we really appreciate this feedback as it makes it clear we can do more to help folks understand how billing works and what their _current_ run rate is.

Let me know if this helps or if I misunderstood your question!

**Comment by jkitchin** [comment]
Yes, that is the storage number I was looking at. Thanks for the clarification.  I guess it is like paying for electricity, e.g. a kilowatt-hr.

Is this GB of stuff in the codespace, including the Docker image, for example?  Do you know if there is a way to tell how big this is?

**Comment by zo0o0ot** [comment]
I'm also confused about this.  When I downloaded the report, a pretty significant amount of the storage was being used by codespaces that I haven't run in months, so I deleted them.  Does that mean that they will not use storage in the future, even though it ruined my storage usage this month and maxed out my storage this month?

**Reply by zo0o0ot** [reply]
Here's my overview- 
![image](https://user-images.githubusercontent.com/876146/202236695-cf4417d1-ac22-41ff-b24f-3f1f9ff4085d.png)
Very little usage, storage absolutely maxed out.  
I requested the detailed report, and found the report had codespaces that I haven't used in months.
I tried deleting codespaces twice (when I got my 75% full warning, and my 90% full warning email messages), expecting the number to go down, and it never did.  I only found out that the storage amount will never go down by finding this discussion.

**Reply by oyetanishq** [reply]
same with me.
I've deleted all my codespaces but still the number is going up????

**Reply by zo0o0ot** [reply]
Yeah, I'm locked out of codespaces unless until December 1st, unless I add a credit card to my account.  Right now, I don't feel comfortable with that because I don't understand how these GB of storage are tracked and whether I have the ability to track what gets built.

Since I had codespaces during the beta, I opened tons of different repos in it to see if I could do anything with it, and sometimes that included attempting to personalize the environment to see if anything would run at all (sometimes successfully, but not always).  Now, I'm seeing the ghosts of those failed experiments coming back to haunt me.

**Reply by oyetanishq** [reply]
😂

**Reply by lizrice** [reply]
I'm also seeing the same thing - haven't used codespaces in ages but suddenly getting reports that I've used all my storage allowance

**Reply by oyetanishq** [reply]
![image](https://user-images.githubusercontent.com/76192403/202619794-5c35ea9b-691e-4326-af36-50e6f7d4c561.png)
There you go.
Got it all exhausted. with no codespace at all.

**Reply by guetler** [reply]
same for me - 16 gb yesterday, today 18.45... increasing - deleted all my unused codespaces.
![Screenshot from 2022-11-19 19-29-31](https://user-images.githubusercontent.com/15184389/202866160-41fb41f5-ddb6-4ce9-ac5e-b9f1ecde9aa6.png)

**Comment by Tnze** [comment]
Why is the Storage increasing even while I haven't using it for months?

**Reply by yangsfang** [reply]
Because you're renting GitHub's datacenter to store your inactive working directory. Suppose you made a change in your work directory, that change by Workspaces' design would persist until the next time you start the Codespace. That data needs to be written and stored on a hard drive purchased by GitHub which also consumes electricity. It is only reasonable you have to pay GitHub for the storage you use for your Codespace, just like all cloud providers such as AWS and Azure charge for their storage products.

However, the problem isn't that you have to pay. Rather, GitHub billing was never clear about at what rate the storage quota was consumed. GitHub Codespace machines are not flexible, they are preset with 32GB storage each. I am speculating your Codespace, whether active or inactive, will consume 0.044444GB of quota each hour, for a total consumption of 32GB over 720 hours, or 30 days. But this is just me speculating based on what would make sense, rather than a clear communication from GitHub. 

It shouldn't hurt GitHub to document that each 32GB storage Codespace should consumer 0.0444444 GB/hr. But it doesn't, because GitHub Codespaces is still new, and the documentation is incomplete.

What would also be awesome is if GitHub Codespaces can be configured with 15GB of storage. Then you can keep a single Codespace for an entire month without having to delete it after each use. Sometimes you do not need 32GB to store your small repository.  This is assuming the kernel, VSCode, and other necessary binaries are stored on a static image shared across all Codespace machines, and thus not counted towards the 32GB quota.

**Comment by zo0o0ot** [comment]
I got this email from GitHub two days ago, which explains a few things:

-------
Codespaces Customer,

Since we launched [Codespaces for All including free usage](https://app.github.media/e/er?s=88570519&lid=3930&elqTrackId=b56b9bca414e4aa981e29ef79b188ab9&elq=f88342ecae4546eaad731d3036d87ae3&elqaid=2898&elqat=1) and [ended the beta for individuals](https://app.github.media/e/er?s=88570519&lid=3929&elqTrackId=57ad240625f7470b898392b2383f3008&elq=f88342ecae4546eaad731d3036d87ae3&elqaid=2898&elqat=1), we have noticed that many of you have old codespaces that were created prior to the availability of [retention period](https://app.github.media/e/er?s=88570519&lid=3928&elqTrackId=10d850741fa84f3da02fc9b0b12854e4&elq=f88342ecae4546eaad731d3036d87ae3&elqaid=2898&elqat=1)s on 22 Jun, 2022 and thus do not have a retention period applied.

In order to help you manage your storage consumption below the free usage allotment, on 21 Nov, 2022 we will apply a retention period to your older codespaces with an expiration of 5 Dec, 2022. If you do nothing, those old codespaces will be deleted on 5 Dec. When you use a codespace, the retention period is restarted, so you don't have to worry about actively used codespaces being deleted.

Codespaces consume storage while they exist - for more information see "[About billing for GitHub Codespaces.](https://app.github.media/e/er?s=88570519&lid=3927&elqTrackId=f6bf32ab43504773872c213e7fa94c0c&elq=f88342ecae4546eaad731d3036d87ae3&elqaid=2898&elqat=1)" Because these older repositories didn't have a retention policy, even if you haven't used it it a long time, your storage is being consumed by these old codespaces. You can see a list of your codespaces at [https://github.com/codespaces](https://github.com/codespaces?elqTrackId=9268d743d2d84105b5bb79af158af2ea&elq=f88342ecae4546eaad731d3036d87ae3&elqaid=2898&elqat=1&elqCampaignId=2842) - a link you can now find quickly from the top of most GitHub.com pages.

Because Codespaces are ephemeral and can be quickly spun up for your work, you can safely delete them once you commit the changes.

Thank you for your use of Codespaces.

Best Regards,
Codespaces Team
-------

So, apparently codespaces now have a default retention period, and you should probably watch that closely if you're on the free tier and don't want storage to run out long before your hours do.

**Reply by oyetanishq** [reply]
I didn't got that much??
Is that mean we are paying for those old codespaces rn?? (I was in beta too).

But I deleted all the codespaces I had??
🙂🙂

**Reply by oyetanishq** [reply]
There is an update!!

I talked to github's support and we found out that there was a bug from github side in which the pre-build of codespaces still work even after deleting the repo (which was happening in my case).

Also that issue was raised to codespaces team and there was a fix rolling out today!!.

**Comment by ghost** [comment]
I didn't use betas and my storage should of been resetted by now because it says 0 days
